File: engine/src/main/java/top/skyeyefast/mchjong/engine/TrainingBot.java
package top.skyeyefast.mchjong.engine;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;

import static top.skyeyefast.mchjong.engine.Action.Type.*;

/** Deterministic decisions from own/public information and engine-issued legal actions. */
final class TrainingBot {
    private final TableView view;
    private final BotDifficulty level;
    private final BotAnalysis analysis;
    private final BotDefence defence;
    private final BotAnalysis.State initial;

    private TrainingBot(TableView view, BotDifficulty level) {
        this.view = view; this.level = level;
        analysis = new BotAnalysis(view, level); defence = analysis.defence; initial = analysis.initial();
    }
    static int choose(TableView view, BotDifficulty level) {
        if (view.actions().isEmpty()) throw new IllegalArgumentException("A bot needs a legal decision");
        for (var type : List.of(RON, TSUMO, NEXT, READY, DRAW_WIND, SHUFFLE, BUILD_WALL, TAKE_PACKET, DRAW)) {
            int index = index(view.actions(), type);
            if (index >= 0) return index;
        }
        if (view.actions().size() == 1) return 0;
        return new TrainingBot(view, level).choose();
    }
    private static int index(List<Action> actions, Action.Type type) {
        for (int i = 0; i < actions.size(); i++) if (actions.get(i).type() == type) return i;
        return -1;
    }
    private record Choice(int index, BotAnalysis.State state, BotAnalysis.Evaluation evaluation,
                          int discard, boolean replacement, double adjustment, String key) {}
    private Choice choice(int index, BotAnalysis.State state, TileEfficiency shape, int discard, boolean replacement, double adjustment) {
        return new Choice(index, state, analysis.evaluate(state, shape, analysis.unseen), discard, replacement, adjustment,
            stable(view.actions().get(index)) + ":" + (discard < 0 ? -1 : BotAnalysis.face(discard)));
    }
    private int choose() {
        var choices = new ArrayList<Choice>();
        var unique = new HashSet<String>();
        var shapes = initial.hand().size() % 3 == 2 ? analysis.discards(initial) : java.util.Map.<Integer, TileEfficiency>of();
        var indices = java.util.stream.IntStream.range(0, view.actions().size()).boxed().sorted(
            Comparator.<Integer, String>comparing(i -> stable(view.actions().get(i)))
                .thenComparing(i -> view.actions().get(i).tiles().stream().sorted().toList().toString())).toList();
        for (int i : indices) {
            var action = view.actions().get(i);
            if (!unique.add(stable(action))) continue;
            switch (action.type()) {
                case DISCARD, RIICHI -> {
                    int tile = action.tiles().getFirst();
                    var next = initial.discard(tile, action.type() == RIICHI);
                    choices.add(choice(i, next, shapes.get(Tile.kind(tile)), tile, false, 0));
                }
                case PASS -> choices.add(choice(i, initial, analysis.shape(initial), Tile.ABSENT, false, 0));
                case CHI, PON -> addCall(choices, i, action);
                case OPEN_KAN, CLOSED_KAN, ADDED_KAN, NUKI -> {
                    var next = declaration(action);
                    var shape = analysis.shape(next);
                    double cost = declarationCost(action, next);
                    double opportunity = Math.min(8, BotAnalysis.live(shape.improving(), analysis.unseen) * .25);
                    choices.add(choice(i, next, shape, Tile.ABSENT, true, opportunity - cost));
                }
                default -> { }
            }
        }
        if (choices.isEmpty()) throw new IllegalStateException("No evaluated legal bot action");
        var baseline = choices.stream().filter(c -> view.actions().get(c.index).type() == DISCARD
            || view.actions().get(c.index).type() == PASS).min(Comparator.<Choice>comparingInt(c -> c.evaluation.shanten())
                .thenComparing(Comparator.comparingDouble((Choice c) -> c.evaluation.utility()).reversed())
                .thenComparing(Choice::key)).orElse(choices.getFirst());
        int minimum = choices.stream().filter(c -> view.actions().get(c.index).type() == DISCARD
            || view.actions().get(c.index).type() == PASS).mapToInt(c -> c.evaluation.shanten()).min().orElse(baseline.evaluation.shanten());
        var mode = defence.mode(baseline.evaluation);
        int abort = index(view.actions(), ABORT_NINE);
        if (abort >= 0 && baseline.evaluation.shanten() >= 4 && baseline.evaluation.live() < 18) return abort;
        Choice fold = null;
        if (mode == BotDefence.Mode.FOLD && !initial.riichi()) {
            int pass = index(view.actions(), PASS);
            // Safety can break completed groups and increase shanten.
            fold = pass >= 0 ? baseline : choices.stream().filter(c -> view.actions().get(c.index).type() == DISCARD)
                .min(Comparator.<Choice>comparingDouble(c -> defence.danger(c.discard))
                    .thenComparing(Comparator.comparingDouble((Choice c) -> c.evaluation.utility()).reversed())
                    .thenComparing(Choice::key)).orElse(baseline);
            var safe = fold;
            // Judge a call by the resulting hand. A valuable, fast continuation
            // can justify attacking even when the unchanged hand would fold.
            choices.removeIf(c -> c != safe && (!viable(c) || defence.mode(c.evaluation) != BotDefence.Mode.PUSH));
            if (choices.size() == 1) return safe.index;
        }
        choices.sort(Comparator.<Choice>comparingDouble(this::score).reversed().thenComparing(Choice::key));
        double bestScore = Double.NEGATIVE_INFINITY;
        Choice best = baseline;
        int roots = 0;
        for (var candidate : choices) {
            var type = view.actions().get(candidate.index).type();
            if ((type == CHI || type == PON || type == OPEN_KAN) && !viable(candidate)) continue;
            if (type == RIICHI && candidate.evaluation.waits().quality() == 0) continue;
            if (candidate != fold && candidate.evaluation.shanten() > minimum + 1) continue;
            // A larger raw ukeire count is not evidence that going backwards is
            // faster. Basic evaluators preserve an available viable route; HARD
            // must actually search a retreat, while dead/yakuless routes can escape.
            boolean safer = candidate.discard >= 0 && baseline.discard >= 0
                && defence.danger(candidate.discard) < defence.danger(baseline.discard);
            boolean retreat = candidate.evaluation.shanten() > minimum && !candidate.replacement && candidate != fold && !safer;
            if (retreat && viable(baseline) && baseline.evaluation.live() > 0
                && (level != BotDifficulty.HARD || roots >= BotAnalysis.SEARCH_ROOTS)) continue;
            double score = score(candidate);
            boolean expand = candidate != fold && (level == BotDifficulty.HARD || candidate.replacement);
            if (expand && roots < BotAnalysis.SEARCH_ROOTS
                && candidate.evaluation.shanten() <= minimum + 1) {
                double forward = analysis.forward(candidate.state, candidate.evaluation, candidate.replacement);
                score += forward - candidate.evaluation.utility() - (candidate.replacement ? Math.min(8, candidate.evaluation.live() * .25) : 0);
                roots++;
            }
            if (score > bestScore || score == bestScore && candidate.key.compareTo(best.key) < 0) {
                bestScore = score; best = candidate;
            }
        }
        return best.index;
    }
    private boolean viable(Choice candidate) {
        return candidate.evaluation.shanten() == 0 ? candidate.evaluation.waits().quality() > 0
            : analysis.value.potential(candidate.state).viable();
    }
    private double score(Choice candidate) {
        double score = candidate.evaluation.utility() + candidate.adjustment;
        if (candidate.discard >= 0) score -= defence.penalty(candidate.discard, defence.mode(candidate.evaluation));
        score += defence.reserve(candidate.state);
        var type = view.actions().get(candidate.index).type();
        if (type == RIICHI) score -= analysis.riichiCost(candidate.evaluation)
            + (defence.placementUrgency(candidate.evaluation.points()) < 1 ? 5 : 0);
        if (type == CHI || type == PON || type == OPEN_KAN) {
            // A closed hand retains a future riichi/tsumo path and defensive options.
            if (initial.melds().stream().allMatch(Meld::closed)) score -= level == BotDifficulty.EASY ? 4 : 8;
            score -= defence.pressure() * 2;
        }
        if (candidate.evaluation.shanten() == 0)
            score += Math.min(8, candidate.evaluation.waits().quality()) * view.riichiSticks() * .4;
        return score;
    }
    private void addCall(List<Choice> choices, int index, Action action) {
        var next = declaration(action);
        var forbidden = LegalActions.forbiddenAfterCall(action, view.focus().tile());
        var shapes = analysis.discards(next);
        var faces = new HashSet<Integer>();
        for (int tile : next.hand()) {
            if (forbidden.contains(Tile.kind(tile)) || !faces.add(BotAnalysis.face(tile))) continue;
            choices.add(choice(index, next.discard(tile, false), shapes.get(Tile.kind(tile)), tile, false, 0));
        }
    }
    private BotAnalysis.State declaration(Action action) {
        var hand = new ArrayList<>(initial.hand());
        for (int tile : action.tiles()) hand.remove(Integer.valueOf(tile));
        var melds = new ArrayList<>(initial.melds());
        var norths = new ArrayList<>(initial.norths());
        switch (action.type()) {
            case NUKI -> norths.addAll(action.tiles());
            case ADDED_KAN -> {
                for (int i = 0; i < melds.size(); i++) {
                    var old = melds.get(i);
                    if (old.type() == Meld.Type.PON && old.kind() == Tile.kind(action.tiles().getFirst())) {
                        var tiles = new ArrayList<>(old.tiles()); tiles.addAll(action.tiles());
                        melds.set(i, new Meld(Meld.Type.ADDED_KAN, tiles, old.fromSeat(), old.calledTile()));
                        break;
                    }
                }
            }
            default -> {
                boolean closed = action.type() == CLOSED_KAN;
                var tiles = new ArrayList<>(action.tiles());
                if (!closed) tiles.add(view.focus().tile());
                tiles.sort(Tile.ORDER);
                melds.add(new Meld(Meld.Type.valueOf(action.type().name()), tiles,
                    closed ? view.viewerSeat() : view.focus().seat(), closed ? Tile.ABSENT : view.focus().tile()));
            }
        }
        boolean blocked = initial.ronBlocked() && (initial.riichi() || !view.rules().callsClearFuriten());
        return new BotAnalysis.State(hand, melds, norths, initial.river(), initial.riichi(), blocked,
            initial.riichi() ? initial.riichiHan() : 1);
    }
    private double declarationCost(Action action, BotAnalysis.State after) {
        int tile = action.tiles().getFirst();
        boolean nuki = action.type() == NUKI;
        double risk = action.type() == OPEN_KAN ? 0 : defence.danger(tile);
        if (action.type() == CLOSED_KAN) risk *= view.rules().robConcealedKan() && Tile.terminalOrHonor(Tile.kind(tile)) ? .08 : 0;
        if (nuki && !view.rules().robNorthWithoutKokushi()) risk *= .08;
        double cost = risk * 15 + (nuki ? 0 : defence.pressure() * 3);
        if (!nuki && view.rules().kanDora()) cost += newIndicatorCost(after);
        return cost;
    }
    private double newIndicatorCost(BotAnalysis.State after) {
        // Unknown indicators are averaged over unseen faces, never revealed in advance.
        int total = java.util.Arrays.stream(analysis.unseen).sum();
        double own = 0, opponents = 0;
        for (int face = 0; face < analysis.unseen.length; face++) {
            if (analysis.unseen[face] == 0) continue;
            int kind = Tile.doraAfter(face % 34, view.rules().sanma());
            double weight = analysis.unseen[face] / (double) Math.max(1, total);
            own += weight * (after.hand().stream().filter(t -> Tile.kind(t) == kind).count()
                + after.melds().stream().flatMap(m -> m.tiles().stream()).filter(t -> Tile.kind(t) == kind).count()
                + after.norths().stream().filter(t -> Tile.kind(t) == kind).count());
            for (var threat : defence.threats) {
                var opponent = view.seats().get(threat.seat());
                double publicCount = opponent.melds().stream().flatMap(m -> m.tiles().stream()).filter(t -> Tile.kind(t) == kind).count()
                    + opponent.norths().stream().filter(t -> Tile.kind(t) == kind).count();
                double concealed = opponent.hand().size() * (analysis.unseen[kind] + analysis.unseen[kind + 34]) / (double) Math.max(1, total);
                opponents += weight * threat.pressure() * (publicCount + concealed);
            }
        }
        return opponents * 5 - own * 3;
    }
    private static String stable(Action action) {
        return action.type().name() + action.tiles().stream().map(BotAnalysis::face).sorted().toList();
    }
}

File: engine/src/main/kotlin/top/skyeyefast/mchjong/engine/TrainingBot.kt
package top.skyeyefast.mchjong.engine

import top.skyeyefast.mchjong.engine.Action.Type.ABORT_NINE
import top.skyeyefast.mchjong.engine.Action.Type.ADDED_KAN
import top.skyeyefast.mchjong.engine.Action.Type.BUILD_WALL
import top.skyeyefast.mchjong.engine.Action.Type.CHI
import top.skyeyefast.mchjong.engine.Action.Type.CLOSED_KAN
import top.skyeyefast.mchjong.engine.Action.Type.DISCARD
import top.skyeyefast.mchjong.engine.Action.Type.DRAW
import top.skyeyefast.mchjong.engine.Action.Type.DRAW_WIND
import top.skyeyefast.mchjong.engine.Action.Type.NEXT
import top.skyeyefast.mchjong.engine.Action.Type.NUKI
import top.skyeyefast.mchjong.engine.Action.Type.OPEN_KAN
import top.skyeyefast.mchjong.engine.Action.Type.PASS
import top.skyeyefast.mchjong.engine.Action.Type.PICK_UP_DICE
import top.skyeyefast.mchjong.engine.Action.Type.ROLL_DICE
import top.skyeyefast.mchjong.engine.Action.Type.PON
import top.skyeyefast.mchjong.engine.Action.Type.READY
import top.skyeyefast.mchjong.engine.Action.Type.RIICHI
import top.skyeyefast.mchjong.engine.Action.Type.RON
import top.skyeyefast.mchjong.engine.Action.Type.SHUFFLE
import top.skyeyefast.mchjong.engine.Action.Type.SKIP_SETTLEMENT
import top.skyeyefast.mchjong.engine.Action.Type.TAKE_PACKET
import top.skyeyefast.mchjong.engine.Action.Type.TSUMO

/** Deterministic decisions from own/public information and engine-issued legal actions. */
internal class TrainingBot private constructor(
    private val view: TableView,
    private val level: BotDifficulty,
    private val diagnostics: MutableList<Diagnostic>? = null,
) {
    private val analysis = BotAnalysis(view, level)
    private val defence = analysis.defence
    private val initial = analysis.initial()

    @JvmRecord
    data class Adjustments(val action: Double, val danger: Double, val reserve: Double, val riichi: Double,
                           val callPressure: Double, val sticks: Double) {
        fun total(): Double = action - danger + reserve - riichi - callPressure + sticks
    }

    @JvmRecord
    data class Diagnostic(val index: Int, val discard: Int, val evaluation: BotAnalysis.Evaluation,
                          val adjustments: Adjustments, val forward: Double, val utility: Double,
                          val search: String, val exclusion: String)

    private data class Choice(
        val index: Int,
        val state: BotAnalysis.State,
        val evaluation: BotAnalysis.Evaluation,
        val discard: Int,
        val replacement: Boolean,
        val adjustment: Double,
        val key: String,
    )

    private fun choice(
        index: Int,
        state: BotAnalysis.State,
        shape: TileEfficiency,
        discard: Int,
        replacement: Boolean,
        adjustment: Double,
    ): Choice = Choice(
        index,
        state,
        analysis.evaluate(state, shape, analysis.unseen),
        discard,
        replacement,
        adjustment,
        stable(view.actions()[index]) + ":" + if (discard < 0) -1 else BotAnalysis.face(discard),
    )

    private fun choose(): Int {
        val choices = mutableListOf<Choice>()
        val unique = hashSetOf<String>()
        val shapes = if (initial.hand().size % 3 == 2) analysis.discards(initial) else emptyMap()
        val indices = view.actions().indices.sortedWith(
            compareBy<Int> { stable(view.actions()[it]) }
                .thenBy { view.actions()[it].tiles().sorted().toString() },
        )
        for (i in indices) {
            val action = view.actions()[i]
            if (!unique.add(stable(action))) continue
            when (action.type()) {
                DISCARD, RIICHI -> {
                    val tile = action.tiles().first()
                    val next = initial.discard(tile, action.type() == RIICHI)
                    choices += choice(i, next, shapes[Tile.kind(tile)]!!, tile, false, 0.0)
                }
                PASS -> choices += choice(i, initial, analysis.shape(initial), Tile.ABSENT, false, 0.0)
                CHI, PON -> addCall(choices, i, action)
                OPEN_KAN, CLOSED_KAN, ADDED_KAN, NUKI -> {
                    val next = declaration(action)
                    val shape = analysis.shape(next)
                    val cost = declarationCost(action, next)
                    val opportunity = minOf(8.0, BotAnalysis.live(shape.improving, analysis.unseen) * 0.25)
                    choices += choice(i, next, shape, Tile.ABSENT, true, opportunity - cost)
                }
                else -> Unit
            }
        }
        if (choices.isEmpty()) throw IllegalStateException("No evaluated legal bot action")
        val baseline = choices.asSequence()
            .filter { view.actions()[it.index].type() == DISCARD || view.actions()[it.index].type() == PASS }
            .minWithOrNull(
                compareBy<Choice> { it.evaluation.shanten }
                    .thenByDescending { it.evaluation.utility }
                    .thenBy { it.key },
            ) ?: choices.first()
        val minimum = choices.asSequence()
            .filter { view.actions()[it.index].type() == DISCARD || view.actions()[it.index].type() == PASS }
            .minOfOrNull { it.evaluation.shanten } ?: baseline.evaluation.shanten
        val mode = defence.mode(baseline.evaluation)
        val abort = index(view.actions(), ABORT_NINE)
        if (abort >= 0 && baseline.evaluation.shanten >= 4 && baseline.evaluation.live < 18) return abort
        var fold: Choice? = null
        if (mode == BotDefence.Mode.FOLD && !initial.riichi()) {
            val pass = index(view.actions(), PASS)
            // Safety can break completed groups and increase shanten.
            fold = if (pass >= 0) {
                baseline
            } else {
                choices.asSequence()
                    .filter { view.actions()[it.index].type() == DISCARD }
                    .minWithOrNull(compareBy<Choice> { defence.danger(it.discard) }.thenByDescending { it.evaluation.utility }.thenBy { it.key })
                    ?: baseline
            }
            val safe = fold
            // Judge a call by the resulting hand. A valuable, fast continuation
            // can justify attacking even when the unchanged hand would fold.
            choices.removeIf {
                val remove = it !== safe && (!viable(it) || defence.mode(it.evaluation) != BotDefence.Mode.PUSH)
                if (remove) record(it, exclusion = "fold")
                remove
            }
            if (choices.size == 1) { record(safe); return safe.index }
        }
        choices.sortWith(compareByDescending<Choice> { score(it) }.thenBy { it.key })
        var bestScore = Double.NEGATIVE_INFINITY
        var best = baseline
        var roots = 0
        for (candidate in choices) {
            val type = view.actions()[candidate.index].type()
            if ((type == CHI || type == PON || type == OPEN_KAN) && !viable(candidate)) {
                record(candidate, exclusion = "no-yaku-route"); continue
            }
            if (type == RIICHI && candidate.evaluation.waits.quality() == 0.0) {
                record(candidate, exclusion = "no-legal-wait"); continue
            }
            if (candidate !== fold && candidate.evaluation.shanten > minimum + 1) {
                record(candidate, exclusion = "shanten"); continue
            }
            // A larger raw ukeire count is not evidence that going backwards is
            // faster. Basic evaluators preserve an available viable route; HARD
            // must actually search a retreat, while dead/yakuless routes can escape.
            val safer = candidate.discard >= 0 && baseline.discard >= 0 &&
                defence.danger(candidate.discard) < defence.danger(baseline.discard)
            val retreat = candidate.evaluation.shanten > minimum && !candidate.replacement && candidate !== fold && !safer
            val exact = level == BotDifficulty.HARD && candidate.evaluation.shanten == 1 && !candidate.replacement
            if (
                retreat && viable(baseline) && baseline.evaluation.live > 0 &&
                (level != BotDifficulty.HARD || !exact && roots >= BotAnalysis.SEARCH_ROOTS)
            ) { record(candidate, exclusion = "unsearched-retreat"); continue }
            var candidateScore = score(candidate)
            var delta = 0.0
            var search = "static"
            val expand = candidate !== fold && (level == BotDifficulty.HARD || candidate.replacement)
            if (expand && (exact || roots < BotAnalysis.SEARCH_ROOTS) && candidate.evaluation.shanten <= minimum + 1) {
                val forward = analysis.forward(candidate.state, candidate.evaluation, candidate.replacement)
                delta = forward - candidate.evaluation.utility -
                    if (candidate.replacement) minOf(8.0, candidate.evaluation.live * 0.25) else 0.0
                candidateScore += delta
                search = if (exact) "one-shanten-exact" else "bounded"
                if (!exact) roots++
            }
            record(candidate, delta, search)
            if (candidateScore > bestScore || candidateScore == bestScore && candidate.key < best.key) {
                bestScore = candidateScore
                best = candidate
            }
        }
        return best.index
    }

    private fun viable(candidate: Choice): Boolean =
        if (candidate.evaluation.shanten == 0) candidate.evaluation.waits.quality() > 0 else candidate.evaluation.potential.viable

    private fun score(candidate: Choice): Double = candidate.evaluation.utility + adjustments(candidate).total()

    private fun adjustments(candidate: Choice): Adjustments {
        val type = view.actions()[candidate.index].type()
        // PASS and the post-call discard already compare speed, live stock,
        // routes, payments and the conditional closed option. Only public threat
        // pressure remains here, rather than charging a fixed opening tax again.
        return Adjustments(candidate.adjustment,
            if (candidate.discard >= 0) defence.penalty(candidate.discard, defence.mode(candidate.evaluation)) else 0.0,
            defence.reserve(candidate.state),
            if (type == RIICHI) analysis.riichiCost(candidate.evaluation) +
                if (defence.placementUrgency(candidate.evaluation.points) < 1) 5 else 0 else 0.0,
            if (type == CHI || type == PON || type == OPEN_KAN) defence.pressure() * 2 else 0.0,
            if (candidate.evaluation.shanten == 0) minOf(8.0, candidate.evaluation.waits.quality()) * view.riichiSticks() * 0.4 else 0.0)
    }

    private fun record(candidate: Choice, delta: Double = 0.0, search: String = "static", exclusion: String = "") {
        if (diagnostics == null) return
        val adjustment = adjustments(candidate)
        diagnostics += Diagnostic(candidate.index, candidate.discard, candidate.evaluation, adjustment, delta,
            candidate.evaluation.utility + adjustment.total() + delta, search, exclusion)
    }

    private fun addCall(choices: MutableList<Choice>, index: Int, action: Action) {
        val next = declaration(action)
        val forbidden = LegalActions.forbiddenAfterCall(action, view.focus().tile())
        val shapes = analysis.discards(next)
        val faces = hashSetOf<Int>()
        for (tile in next.hand()) {
            if (Tile.kind(tile) in forbidden || !faces.add(BotAnalysis.face(tile))) continue
            choices += choice(index, next.discard(tile, false), shapes[Tile.kind(tile)]!!, tile, false, 0.0)
        }
    }

    private fun declaration(action: Action): BotAnalysis.State {
        val hand = initial.hand().toMutableList()
        for (tile in action.tiles()) hand.remove(tile)
        val melds = initial.melds().toMutableList()
        val norths = initial.norths().toMutableList()
        when (action.type()) {
            NUKI -> norths.addAll(action.tiles())
            ADDED_KAN -> {
                for (i in melds.indices) {
                    val old = melds[i]
                    if (old.type() == Meld.Type.PON && old.kind() == Tile.kind(action.tiles().first())) {
                        val tiles = old.tiles().toMutableList()
                        tiles.addAll(action.tiles())
                        melds[i] = Meld(Meld.Type.ADDED_KAN, tiles, old.fromSeat(), old.calledTile())
                        break
                    }
                }
            }
            else -> {
                val closed = action.type() == CLOSED_KAN
                val tiles = action.tiles().toMutableList()
                if (!closed) tiles += view.focus().tile()
                tiles.sortWith(Tile.ORDER)
                melds += Meld(
                    Meld.Type.valueOf(action.type().name),
                    tiles,
                    if (closed) view.viewerSeat() else view.focus().seat(),
                    if (closed) Tile.ABSENT else view.focus().tile(),
                )
            }
        }
        val blocked = initial.ronBlocked() && (initial.riichi() || !view.rules().callsClearFuriten())
        return BotAnalysis.State(
            hand,
            melds,
            norths,
            initial.river(),
            initial.riichi(),
            blocked,
            if (initial.riichi()) initial.riichiHan() else 1,
        )
    }

    private fun declarationCost(action: Action, after: BotAnalysis.State): Double {
        val tile = action.tiles().first()
        val nuki = action.type() == NUKI
        var risk = if (action.type() == OPEN_KAN) 0.0 else defence.danger(tile)
        if (action.type() == CLOSED_KAN) {
            risk *= if (view.rules().robConcealedKan() && Tile.terminalOrHonor(Tile.kind(tile))) 0.08 else 0.0
        }
        if (nuki && !view.rules().robNorthWithoutKokushi()) risk *= 0.08
        var cost = risk * 15 + if (nuki) 0.0 else defence.pressure() * 3
        if (!nuki && view.rules().kanDora()) cost += newIndicatorCost(after)
        return cost
    }

    private fun newIndicatorCost(after: BotAnalysis.State): Double {
        // Unknown indicators are averaged over unseen faces, never revealed in advance.
        val total = analysis.unseen.sum()
        var own = 0.0
        var opponents = 0.0
        for (face in analysis.unseen.indices) {
            if (analysis.unseen[face] == 0) continue
            val kind = Tile.doraAfter(face % 34, view.rules().sanma())
            val weight = analysis.unseen[face] / maxOf(1, total).toDouble()
            own += weight * (
                after.hand().count { Tile.kind(it) == kind } +
                    after.melds().sumOf { meld -> meld.tiles().count { Tile.kind(it) == kind } } +
                    after.norths().count { Tile.kind(it) == kind }
                )
            for (threat in defence.threats) {
                val opponent = view.seats()[threat.seat]
                val publicCount = opponent.melds().sumOf { meld -> meld.tiles().count { Tile.kind(it) == kind } } +
                    opponent.norths().count { Tile.kind(it) == kind }
                val concealed = opponent.hand().size * (analysis.unseen[kind] + analysis.unseen[kind + 34]) /
                    maxOf(1, total).toDouble()
                opponents += weight * threat.pressure * (publicCount + concealed)
            }
        }
        return opponents * 5 - own * 3
    }

    companion object {
        /** Opt-in trace uses exactly the same candidate generation and search as play. */
        @JvmStatic
        fun inspect(view: TableView, level: BotDifficulty): List<Diagnostic> {
            if (view.actions().none { it.type() in listOf(DISCARD, RIICHI, PASS, CHI, PON, OPEN_KAN, CLOSED_KAN, ADDED_KAN, NUKI) }) return emptyList()
            val diagnostics = mutableListOf<Diagnostic>()
            TrainingBot(view, level, diagnostics).choose()
            return diagnostics
        }

        @JvmStatic
        fun choose(view: TableView, level: BotDifficulty): Int {
            if (view.actions().isEmpty()) throw IllegalArgumentException("A bot needs a legal decision")
            for (type in listOf(RON, TSUMO, SKIP_SETTLEMENT, NEXT, READY, DRAW_WIND, SHUFFLE, BUILD_WALL, PICK_UP_DICE, ROLL_DICE, TAKE_PACKET, DRAW)) {
                val index = index(view.actions(), type)
                if (index >= 0) return index
            }
            if (view.actions().size == 1) return 0
            return TrainingBot(view, level).choose()
        }

        private fun index(actions: List<Action>, type: Action.Type): Int = actions.indexOfFirst { it.type() == type }

        private fun stable(action: Action): String =
            action.type().name + action.tiles().map(BotAnalysis::face).sorted()
    }
}
